src/modules/Currency/Service.php

```diff
@@ -522,50 +484,185 @@ public function updateCurrencyRates($data)
+    {
+        $result = $this->di['cache']->get("exchangerate.api.$from.$key.$validFor", function (ItemInterface $item) use ($from, $validFor, $key) {
+            $from_Currency = urlencode($from);
+
+            if (!empty($key)) {
+                $key = urlencode($key);
+                $requestUrl = "https://v6.exchangerate-api.com/v6/$key/latest/$from_Currency";
+            } else {
+                $requestUrl = 'https://open.er-api.com/v6/latest/USD';
+            }
+
+            $client = HttpClient::create(['bindto' => BIND_TO]);
+            $response = $client->request('GET', $requestUrl);
+            $array = $response->toArray();
+
+            if ($array['result'] !== 'success') {
+                $item->expiresAfter(15 * 60 * 60); // Try again in 15 min
+                error_log('ExchangeRate-API Gave an error: ' . $array['error-type']);
+
+                throw new \FOSSBilling\Exception('There was an error when fetching currency rates from ExchangeRate-API. See the error log for details.');
             }
 
-            throw new \FOSSBilling\Exception('Failed to get currency rates for :currency from the European Central Bank API', [':currency' => $to_Currency]);
+            if ($validFor === 0) {
+                // ExchangeRate-API is great and will tell us exactly when the data will next have an update, so we will use that for the cache expiration when "auto" is the sync mode.
+                $item->expiresAt(new \DateTime($array['time_next_update_utc']));
+            } else {
+                $item->expiresAfter($validFor);
+            }
+
+            return $array;
+        });
+
+        // Their open access API endpoint has a specific param to inform of if it ever goes EOL, so let's monitor that and trigger an error to alert us if it's deprecated
+        if (array_key_exists('time_eol_unix', $result) && $result['time_eol_unix'] !== 0) {
+            trigger_error('ExchangeRate-API has deprecated their open endpoint. Investigate!', E_USER_DEPRECATED); // Should be sent via error reporting, making monitoring this easy
+        }
+
+        // Different array key between the open and authenticated endpoint, but otherwise it's the same structure.
+        if (!empty($key)) {
+            return $result['conversion_rates'] ?? [];
         } else {
+            return $result['rates'] ?? [];
+        }
+    }
+
+    /**
+     * Gets the rates from https://apilayer.com/marketplace/currency_data-api.
+     * Fetches a complete list off currencies and then caches that result for the specified period.
+     * Normalizes the return array.
+     */
+    protected function getCurrencyDataRates(string $from, int $validFor, string $key)
+    {
+        $result = $this->di['cache']->get("currency.data.api.$from.$key.$validFor", function (ItemInterface $item) use ($from, $validFor, $key) {
+            $item->expiresAfter($validFor);
+
+            $from_Currency = urlencode($from);
```

Comment:
Any particular reason for capitalising the second word? 🧐

Reply:
Nope.. just had a moment when typing apparently.
Fixed now :)